### ISSUE TITLE
C grid noise in corners

### ISSUE THREAD

**Issue by JFLemieux73** [CICE-Consortium/CICE#792]
See also issue #791.

**Comment by JFLemieux73** [comment]
There is noise in the thickness and concentration fields in corners with the C grid. Here is again the figure from issue #791:

![Cevp_NE_cdx_P1e4_aice_2005-01-15-00000](https://user-images.githubusercontent.com/31927797/202021409-7b4266bc-cda4-4202-85fb-f5fbc8cb754f.png)

Note that this is obtained with the uniform grid.

**Comment by JFLemieux73** [comment]
Reducing the time step from 60 min to 15 min does not change the pattern:

![Cevp_NE_cdx_P1e4_15min_aice_2005-01-15-00000](https://user-images.githubusercontent.com/31927797/202025086-f797e7d8-a2a5-4cf0-8d65-85d5304e4299.png)

**Comment by JFLemieux73** [comment]
Hum...the noise disappears with upwind:

![Cevp_NE_cdx_P1e4_15min_upwind_aice_2005-01-15-00000](https://user-images.githubusercontent.com/31927797/202030393-631a67ab-1fd1-43c0-b015-ed5d79d8ae85.png)

**Comment by dabail10** [comment]
I do think that the incremental remapping on the B-grid is a problem with the C-grid. We could workaround this by introducing a slip condition along the land boundaries for velocity.

**Comment by eclare108213** [comment]
It looks like a null-space solution is being energized. The checkerboard pattern is in the null space for some operators on B grids, and C grids probably have similar issues.  Upwind advection is very diffusive, enough to keep null-space solutions from growing, but incremental remapping is much less diffusive.  Initially, you could try averaging the C-grid velocities to the B-grid corners and running the remapping in the standard way, without the C-grid changes we made.  That defeats the purpose of developing the C-grid discretization, but it's just a test.  Allowing a slip condition with remapping could help, or just use upwind for the C-grid, for now.  The best solution to this problem might be to rewrite incremental remapping for a C grid. That's been started but needs a funding source.

**Comment by JFLemieux73** [comment]
Thanks both of you for your inputs. Elizabeth, is the null-space issue related to the C grid (momentum) discretization, the remapping (using C grid velocities interpolated to the U points) or a combination of both?

I don't understand your sentence "Initially, you could try averaging the C-grid velocities to the B-grid corners and running the remapping in the standard way, without the C-grid changes we made. "...isn't it what we are doing right now?

**Comment by JFLemieux73** [comment]
Maybe a solution would be to apply upwind only close to land and remapping for the rest of the domain...

**Comment by eclare108213** [comment]
> is the null-space issue related to the C grid (momentum) discretization, the remapping (using C grid velocities interpolated to the U points) or a combination of both?

It's probably a combination, but note that the remapping is not using U velocities - see below.

It's definitely related to the C-grid discretization -- avoiding the B-grid checkerboard instability is the main reason I do 4 calculations for each grid cell (ne, se, sw, nw) in EVP.  I also think it's important to ensure that energy is dissipated in the discretized equations in the same manner as in the continuous equations, which the 4 calculations do, when combined for F1 and F2; I'm not sure whether your discretization guarantees that, but others have been successful with it so there's hope.

> "Initially, you could try averaging the C-grid velocities to the B-grid corners and running the remapping in the standard way, without the C-grid changes we made. "...isn't it what we are doing right now?

No, incremental remapping needs the edge velocities, so for the B grid, the corner velocities are interpolated to the edges.  For the C grid, we took that out and use the edge velocities directly.  Not doing that (i.e. reverting to the B-grid process) would add some diffusion into the solution associated with the averaging and also probably change the behavior near land, due to the masking.

**Comment by JFLemieux73** [comment]
I am confused. I think remapping uses the U velocities directly. See section 2.4 in the doc:

"as the remapping scheme was originally developed for B grid applications,  uvel and vvel are directly used for the advection."

I could check the code but my impression is that for the C grid we interpolate uvelE and vvelN to the U point in order to use the remapping.

**Comment by eclare108213** [comment]
Look at the code here:
https://github.com/CICE-Consortium/CICE/blob/9808b5195b2af77d2b70f5fec3e417b0a6020c5f/cicecore/cicedynB/dynamics/ice_transport_driver.F90#L544
and
https://github.com/CICE-Consortium/CICE/blob/9808b5195b2af77d2b70f5fec3e417b0a6020c5f/cicecore/cicedynB/dynamics/ice_transport_remap.F90#L667

I think the doc needs to be corrected by adding E, N, U, etc to uvel and vvel, to not be so confusing.

**Comment by eclare108213** [comment]
I looked at the doc, and I don't think it's right.  But maybe I'm wrong.  @TillRasmussen please straighten us out on this!

Quote from https://cice-consortium-cice.readthedocs.io/en/main/science_guide/sg_horiztrans.html

> Two transport schemes are available: upwind and the incremental remapping scheme of [[10]](https://cice-consortium-cice.readthedocs.io/en/main/zreferences.html#dukowicz00) as modified for sea ice by [[37]](https://cice-consortium-cice.readthedocs.io/en/main/zreferences.html#lipscomb04). The upwind scheme is naturally suited for a C grid discretization. As such, the C grid velocity components (i.e. _uvelE=u_ at the E point and _vvelN=v_ at the N point) are directly passed to the upwind transport scheme. On the other hand, if the B grid is used, _uvel_ and _vvel_ (respectively _u_ and _v_ at the U point) are interpolated to the E and N points such that the upwind advection can be performed. Conversely, as the remapping scheme was originally developed for B grid applications, _uvel_ and _vvel_ are directly used for the advection. If the remapping scheme is used for the C grid, _uvelE_ and _vvelN_ are first interpolated to the U points before performing the advection.

This is what I think it should say, starting with the 'Conversely' sentence above:

> ... Because the remapping scheme was originally developed for B grid applications, _uvelU_ and _vvelU_ are are interpolated to the E and N points internally in the advection code, when running on B grids. If the remapping scheme is used for the C grid, then this interpolation step is skipped and _uvelE_ and _vvelN_ are used directly.

**Comment by TillRasmussen** [comment]
I think that you are right. I would maybe rephrase it as

Two transport schemes are available: upwind and the incremental remapping scheme of [[10]](https://cice-consortium-cice.readthedocs.io/en/main/zreferences.html#dukowicz00) as modified for sea ice by [[37]](https://cice-consortium-cice.readthedocs.io/en/main/zreferences.html#lipscomb04). Both schemes are naturally suited for a C grid discretization. As such, the C grid velocity components (i.e. uvelE=u at the E point and vvelN=v at the N point) are directly passed to the transport scheme. On the other hand, if the B grid is used, uvel and vvel (respectively u and v at the U point) are interpolated to the E and N points such that the advection can be performed.

**Comment by JFLemieux73** [comment]
Thanks Till.

Elizabeth, you wrote above: "I'm not sure whether your discretization guarantees that, but others have been successful with it so there's hope."

Looking at the overleaf document again (see section 5), our implementation is not exactly what others have done. Again, this is due to the fact that we started with a CD grid. Maybe we should test exactly what Kimmritz had? Basically, we would need a different approach for calculating Ds at the T point.

**Comment by eclare108213** [comment]
Well, I'm also not sure that theirs guarantees it, or that yours is somehow worse than theirs for any reason, or on the other hand that either is bad.  We would need to do some actual numerical analysis on the discretization.  But if it's straightforward to test Kimmritz's approach, then you could try that.  What do the others use for advection?

**Comment by JFLemieux73** [comment]
I have looked at the C grid divu field (at different times) when using remapping. It is indeed noisy. If the noise comes from the solution of the momentum equation, I guess I should be able to see it in instantaneous divu fields even using upwind. I have looked at divu at many different times and don't see any noise. The C and B grid divu fields are very similar when using upwind (not shown).

**Comment by eclare108213** [comment]
Does it only happen with corners?  i.e. if you set up a box case with a wall on one side and periodic in the other direction, would it show the checkerboarding?

**Comment by JFLemieux73** [comment]
Good idea. I will test it.

**Comment by eclare108213** [comment]
You could try with the forcing perpendicular to the wall and also at an angle, e.g. northeastward if the wall is on the east side.

**Comment by JFLemieux73** [comment]
Here is a channel test case with winds blowing north. The north and south boundaries have walls and cyclic condition is used along the x axis (west-east). We don't see the checkerboard pattern with C grid and remap:

![Cperio_remap_aice_2005-01-15-00000](https://user-images.githubusercontent.com/31927797/202796941-476ac275-8ee8-48a0-a4e6-bd92d89fbb84.png)

**Comment by JFLemieux73** [comment]
As suggested by Elizabeth, I kept the same test case but used atm_data_type   = 'uniform_northeast' instead. The conclusion is the same: there is no checkboard pattern (not shown).

**Comment by JFLemieux73** [comment]
Tony I think we have seen this before...I thought we had fixed that problem but I guess I was wrong.

**Comment by JFLemieux73** [comment]
I also tested with avg_strength instead of avg_zeta. It does not fix the problem.

**Comment by JFLemieux73** [comment]
For the other problem there is only eastblock that is available. I will come back later to this problem and focus on the corner problem.

**Comment by JFLemieux73** [comment]
I confirm what Till wrote above. uvelE and vvelN are not directly used as l_fixed_area = .false. If the remap is called as for the B-grid  (uvelE and vvelN are not sent) the answer is the same (BFB) with obviously the same checkerboard pattern.

**Comment by eclare108213** [comment]
Reviewing and catching up on this discussion.  Some thoughts:

> I have looked at the C grid divu field (at different times) when using remapping. It is indeed noisy. If the noise comes from the solution of the momentum equation, I guess I should be able to see it in instantaneous divu fields even using upwind. I have looked at divu at many different times and don't see any noise. The C and B grid divu fields are very similar when using upwind (not shown).

This is consistent with what Martin said, "that a C-grid is accurate for (flux) divergence operations, but can have some noise in the staggered velocities." 

In your eastblock case above, presumably the only difference from the cyclic channel with the wind blowing north, other than the direction, is that there is an ice edge along the wall, i.e. a corner formed from the ice and the wall itself.  Is that correct? I wonder if the CD grid velocities, with appropriate averaging, would be enough to damp the checkerboard instability.

The [CICE documentation for incremental remapping](https://cice-consortium-cice.readthedocs.io/en/main/science_guide/sg_horiztrans.html) includes this intriguing statement:
> We made one other change in the scheme of [10] for locating triangles. In their paper, departure points are defined by projecting cell corner velocities directly backward. ... This approximation is only first-order accurate. Accuracy can be improved by estimating the velocity at the midpoint of the trajectory.

It would be fairly complicated to implement, but could using those velocities interpolated directly from the calculated uvelE, vvelN (without first interpolating them to the corners) bypass the checkerboard pattern associated with the corner velocities?  If the issue stems from the momentum calculation, as Martin asserts, then I suspect this would not help.  The increased accuracy might even make it worse.

Here's another thought: what if you did allow a small amount of free slip along coastlines, as @dabail10 suggested?  ~puny.  I guess it could be implemented by changing the inland value (currently set so that velocity is strictly zero at the coast) slightly, so that the effective coastline moves just a little.  Maybe that's a crazy idea!  Just throwing things out there...

**Comment by JFLemieux73** [comment]
Thanks Elizabeth. I am glad you are throwing ideas as I am a bit confused about what I should do next! :)

"In your eastblock case above, presumably the only difference from the cyclic channel with the wind blowing north, other than the direction, is that there is an ice edge along the wall, i.e. a corner formed from the ice and the wall itself. Is that correct?" 

yes it is correct.

"I wonder if the CD grid velocities, with appropriate averaging, would be enough to damp the checkerboard instability."

I am not ready to go back to the CD grid yet...but let's keep that in mind.

"Here's another thought: what if you did allow a small amount of free slip along coastlines, as @dabail10 suggested? ~puny. I guess it could be implemented by changing the inland value (currently set so that velocity is strictly zero at the coast) slightly, so that the effective coastline moves just a little. Maybe that's a crazy idea! Just throwing things out there..."

Yes I could try something like that. I would like also to try to set u and/or v to zero close to land (kind of extend the BCs).

**Comment by JFLemieux73** [comment]
Another thing I would like to try is to use upwind close to land and remap elsewhere...

**Comment by JFLemieux73** [comment]
Finally for some tests the eastblock is better. The last ocean T cell on the right is T(79,j). In stepv_C, I added:

if (i .eq. 79) then
            vvel(i,j)=0.0
 endif

This kind of extends the no slip condition. The checkerboard is still there (not shown). 

Commenting the above and now adding in stepu_C:

 if (i .eq. 78) then
            uvel(i,j)=0.0
 endif

This kind of extends the no outflow condition. The checkerboard is still there (not shown). 

Including both conditions at the same time (extend no slip, no outflow conditions) also leads to the checkerboard pattern...this suggests that the problem is not related to the boundary conditions.

**Comment by eclare108213** [comment]
This [paper](https://perso.uclouvain.be/emmanuel.hanert/wp-content/plugins/zotpress/lib/request/request.dl.php?api_user_id=5103066&dlkey=CE9Q53VM&content_type=application/pdf) has some potentially useful background info and references.  The [POP manual](https://perso.uclouvain.be/emmanuel.hanert/wp-content/plugins/zotpress/lib/request/request.dl.php?api_user_id=5103066&dlkey=CE9Q53VM&content_type=application/pdf) for CESM also mentions the checkerboard (p 19).

**Comment by JFLemieux73** [comment]
Thanks Elizabeth. I will read it. I also wanted to test upwind at the beginning and then switch to remap. It does not help...the checkerboard develops when remap is switched on. I tested 2 or 5 days of upwind (remap after) with the eastblock test. Results after 15 days show the checkerboard...but less pronounced close to the wall (not shown).

**Comment by JFLemieux73** [comment]
So I configured the McGill (C-grid) model to have similar test cases. I guess I don't have exactly the same parameters (e.g. drag coefficients) but here is a quick test. Here is the corner test case after 15 days with upwind:

![A1990_01_15_00_00_01](https://user-images.githubusercontent.com/31927797/204896272-bad1d685-e5b6-4297-bf7e-8d4b19ed38e9.png)

As with CICE (C-grid) with the upwind we don't see the checkerboard pattern. Maybe to be expected as upwind is diffusive.

**Comment by JFLemieux73** [comment]
I recently coded a semi-lagrangian advection scheme in the McGill model. This is similar to the remapping scheme and a lot less diffusive than upwind. Note that close to land (2 cells) the order of the scheme has to be reduced. Close to land upwind is therefore applied. Here is the concentration after 15 days with the semi-lag scheme:

![A1990_01_15_00_00_02](https://user-images.githubusercontent.com/31927797/204897876-f0e1f24f-c859-4933-be11-7b74bc7d5157.png)

**Comment by JFLemieux73** [comment]
Apart from slightly different parameters, I have the impression that there are two things that are different between the McGill runs and the CICE runs: JFNK versus and EVP and the treatment of the BCs. I really don't think it is a solver issue (JFNK versus EVP). 

Both models have no-slip, no outflow BCs but they are implemented differently: ghost cells for CICE and Taylor series expansions (complicated!!!) for McGill. Even though I wrote above that BCS are not the issue maybe it needs to be reconsidered. 

So at this point I suspect that the checkerboard comes from one or the two following:
1) the implementation of BCs
2) the remapping scheme itself when used for the C-grid.

Elizabeth any ideas?

**Comment by JFLemieux73** [comment]
I have also tested the eastblock. There is no checkerboard in both upwind and semilag runs (not shown).

**Comment by eclare108213** [comment]
I agree that 3) is likely not important for this issue, but I do think it is wrong in the McGill model - see W. M. Connolley, J. M. Gregory, E. C. Hunke and A. J. McLaren (2004). On the consistent scaling of terms in the sea ice dynamics equation. J. Phys. Oceanogr., 34, 1776–1780.

2) should be fixed, but it's likely not impacting the northeast corner case.

That leaves 1)...

**Comment by JFLemieux73** [comment]
I agree. Ok I will start with 1)...

**Comment by JFLemieux73** [comment]
I think with avg_zeta the only thing I need to change for the BCs is dvdx in strain_rates_U. dudy is zero at the wall (eastern wall). The U point at the wall is (i,j). dvdx at Uij can be found with Taylor series expansions. It is expressed as:

dvdx ~ vvelN(i-1,j) / (3 dx) - 3 vvelN(i,j) / dx

Indeed the stencil is larger with this  than with our current (ghost cell) method.

**Comment by JFLemieux73** [comment]
Ok I have added this in strain_rates_U after the calculation of shearU:

if (npm(i,j) == 1 .and. npm(i+1,j) == 0) then
            shearU(i,j) = dyU(i,j) * (vvelN(i-1,j) / 3d0 - 3d0*vvelN(i,j))
endif

I am not available for the rest of the day...we will know tomorrow morning if this works.

**Comment by JFLemieux73** [comment]
It is disappointing but the new BC does not change anything; the checkerboard is still there (not shown).

**Comment by JFLemieux73** [comment]
Elizabeth, I am a bit confused about what I should try next...should I try the ice edge velocities 2)?

**Comment by dabail10** [comment]
Did you already change the subcycles in the EVP? I was looking back at the paper led by Bill Lipscomb:

https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2005JC003355

This is more about high resolution ... However, is there a B/C grid issue with the ridging?

**Comment by JFLemieux73** [comment]
Thanks Dave. I don't think it is an EVP issue. I use  elasticDamp     = 0.1d0 and ndte=2400 (10 times bigger than default). These values of  elasticDamp and ndte should increase the numerical convergence. I know that paper led by Bill. This problem should be less of an issue with kstrength=0 (Hibler...this is what I use) and as dt is reduced. I use 60 min right now and I think I tested 15 min before. Maybe I could give it a try with 5 min. Stay tuned.

**Comment by eclare108213** [comment]
I think implementing the ice edge velocities would be interesting and useful for other purposes, but I do not think that it will help this issue.  I want to dig back into my evp notes to see what the discretization might look like in that approach (bigger stencil).  I've rederived it since the original derivation, so I should be able to reproduce the steps needed, I'm just not sure what it would look like for a C grid.

**Comment by JFLemieux73** [comment]
If we want to implement another discretization we should consider what Bill proposed;

https://github.com/CICE-Consortium/CICE/issues/714

This could be our next coding camp! 😎

**Comment by JFLemieux73** [comment]
Dave,
I ran it with a dt=120 s...the checkerboard is still there.

**Comment by eclare108213** [comment]
Wow!  I'm not familiar with how l_fixed_area is supposed to work, not having ever used it, but my impression is that you need to externally calculate some things to send into the transport routine in addition to changing the flag.  So this might not be actually transporting anything? @whlipscomb might be able to shed some light on what's needed here.

**Comment by dabail10** [comment]
Are you sure you had incremental remapping on?

**Comment by JFLemieux73** [comment]
Yes.  advection  = 'remap'

**Comment by JFLemieux73** [comment]
Here are some comments taken from ice_transport_remap.F90:

! Finally, a few words about the edgearea fields:
!
! In earlier versions of this scheme, the divergence of the velocity
! field implied by the remapping was, in general, different from the
! value of del*u computed in the dynamics.  For energetic consistency
! (in CICE as well as in layered ocean models such as HYPOP),
! these two values should agree.  This can be ensured by setting
! l_fixed_area = T and specifying the area transported across each grid
! cell edge in the arrays edgearea_e and edgearea_n.  The departure
! regions are then tweaked, following an idea by Mats Bentsen, such
! that they have the desired area.  If l_fixed_area = F, these regions
! are not tweaked, and the edgearea arrays are output variables.

**Comment by TillRasmussen** [comment]
Would it be worthwhile to reach out to Mats Bentsen?

**Comment by JFLemieux73** [comment]
We could start with Bill's opinion and then maybe we could contact Mats. In the meantime I am trying to better understand the remapping...

**Comment by eclare108213** [comment]
I am working up a discretization using the variational approach, as in the current B-grid EVP.  I'm not sure it will work, since C-grid doesn't have enough velocity nodes to use a fully bilinear approximation -- this means that the discretization will be lower order than in a bilinear case.  It _might_ be less likely to checkerboard, but there's no guarantee that it won't, and we also might get some insight into how the boundary conditions play in.

**Comment by JFLemieux73** [comment]
Happy new year! I still have to investigate but results above suggest that the checkerboard comes from the remapping not the C grid discretization...do we need the variational approach?

**Comment by eclare108213** [comment]
I think it's important, but I'm willing to entertain other ideas.  My understanding is that the advection scheme is not creating the checkerboard so much as allowing it, because a feature of incremental remapping is its very low diffusion -- a good thing as far as advection goes.  I interpret your results above as advection schemes (or maybe energetic inconsistencies) in the other codes covering up the problems, rather than fixing the source.  On the B-grid, the variational approach eventually led to a discretization that did not checkerboard while using the same advection scheme being used here, and it's not a discretization that could simply be written down based on the form of the continuous equations.  In deriving the current B-grid EVP scheme, I went through several iterations that did checkerboard, even while using the variational approach for the derivation -- those schemes were much simpler (lower order).  The key was to use a higher order basis function (bilinear) for the velocity AND stresses, which is what produces the 4 values per grid cell (nw, ne, sw, se) for all of the components.

The problem I'm having with the C-grid is that it does not have enough degrees of freedom (which Sergei also pointed out, I think).  I can use a bilinear basis function for the stresses but I haven't found a basis that works for velocities located on the edge mid-points.  The discretization I just worked out uses a bilinear stress approximation (on corners) but only linear for the velocities (i.e. assuming u varies in the x direction and is constant in y, similarly for v).  I just reached the end of the derivation and it is not correct for sigma12, which needs du/dy etc. (I should have realized that earlier -- lesson learned).  I could write down what the sigma12 discretization would be based on the forms of sigma1 and sigma2, but then it wouldn't be consistent with the divergence, tension and stress terms. (Unless maybe we somehow change the boundary conditions?)  The variational approach ensures that all of these things are consistent with each other, as they must be to guarantee that the discretized equations dissipate energy in the same manner as the continuous equations.  That's why I think it's important.

**Comment by whlipscomb** [comment]
@JFLemieux73, I'm having to revisit my long-ago youth. But fortunately, I wrote some comments on l_fixed_area in ice_transport_remap.F90.

According to those comments, this tweak ensures that the area fluxed across each cell edge is equal to the area implied by del*u in the dynamics. On a C-grid, u is simply the cell-edge velocity, instead of an average from neighboring corners as on the B-grid.

It would be easier to show with a picture, but I'll try in words.  The idea is to tweak the size of the departure region without too much changing its basic shape. Try drawing a north cell edge with a departure region above.  The departure region is a quadrilateral defined by the two edge vertices and the departure points of the two back trajectories. The velocity in this case would be mainly in the negative-y direction. For simplicity, assume that both departure points lie inside the cell above the edge (i.e., they don't lie in the adjacent cells to the left and right).

On the segment connecting the two departure points, mark the midpoint. From the midpoint, draw a segment to each edge vertex. This divides the departure region into three triangles.

Now slide the midpoint up a little bit. This slightly increases the area of the three triangles. If you slide the midpoint down, the three triangles get smaller.

The way the code works is that it figures out how much to move the midpoint. Otherwise, the midpoint stays put, and the logic is the same as for l_fixed_area = F.

It doesn't surprise me that setting l_fixed_area = T suppresses noise, but I'm not sure I'd have predicted it either. It's been too long since I thought about it. 

But if it works, I'd say to go ahead and use this setting with the C grid. At some later time, we could write a simpler remapping scheme that's more natural on a C-grid. But I'm guessing that for now, you're mainly looking for a fix that works.

Please let me know if this description isn't clear.

**Comment by JFLemieux73** [comment]
The C grid noise has been eliminated with l_fixed_area=T (see issues #809 and #811). The code does not crash anymore for gx simulations (see PR #833).
